Clipboard Support

ricanteja opened the pull request:
Hey guys, it's been a while.

So those that follow me on Twitter know that I've tried my hand at implementing clipboard support for Windows and Linux. Here are the fruits of my labor.

The API is thus:

```
sf::String str = sf::Clipboard::getString();
sf::Clipboard::setString(const sf::String& str);
```

Static functions with no parameter. I got it working on Linux as well using some hidden globals.

Note: I accidentally created another pull request that was all messed up. It can be disregarded and deleted.

eXpl0it3r commented:
Any specific reason you use pthread thread, lock and mutex instead of the SFML equivalents for the Unix implementation?

eXpl0it3r commented on src/SFML/Window/Unix/ClipboardImpl.cpp:
```diff
@@ -0,0 +1,268 @@
+                    {
+                        if(event.xselectionrequest.selection == selection)
+                        {
+                            std::cerr << "Sending clipboard data" << std::endl;
```
I assume this is a left over debug output, right?

ricanteja replied:
Oops yeah, there is still some cleaning up I need to do. Thanks for spotting that.

eXpl0it3r commented:
Also, am I understanding this right, that on Unix, you'll create a thread with a hidden window and that thread will busy wait with a 20ms sleep time? Isn't it possible to wait for events, instead of polling them?
And finally, how does handling the event queue within that thread affect the event queues of other windows?

eXpl0it3r commented on src/SFML/Window/Win32/ClipboardImpl.cpp:
```diff
@@ -0,0 +1,103 @@
+////////////////////////////////////////////////////////////
+String ClipboardImpl::getString()
+{
+	String text;
```
Tab indentation instead of space indentation.

ricanteja commented:
Thanks eXpl0it3r for giving my pr some attention.

I had a weird bug with sf::Thread where the host thread wasn't exiting when the program quit. I'll try to reproduce it later. Using pthread got rid of the issue.

ricanteja commented:
The thread that hosts the clipboard should not effect the event queue of the other windows at all I don't think. From what I've tested there are no issues. It gets events using it's own display and window.

mantognini commented:
> I had a weird bug with sf::Thread where the host thread wasn't exiting when the program quit. I'll try to reproduce it later. Using pthread got rid of the issue.

It's just a guess, but it might be related to this fact about `sf::Thread`:

> If the owner sf::Thread instance is destroyed before the thread is finished, the destructor will wait (see wait())

ricanteja commented:
I think you are right @mantognini ...

I just pushed some new changes to the PR. I removed the extra debug message I had in there and got rid of some headers I wasn't using.

Here is some example code for testing this PR

```cpp
#include <iostream>
#include <SFML/Graphics.hpp>
#include <SFML/Window/Clipboard.hpp>


int main()
{
    sf::RenderWindow window(sf::VideoMode(420, 380), "SFML Clipboard");

    sf::Font font;

    // Replace with your font file
    if(!font.loadFromFile("./YourFont.ttf"))
    {
        std::cerr << "Failed to load font file." << std::endl;
        return -1;
    }

    sf::Text text("Clipboard!", font);


    while(window.isOpen())
    {
        sf::Event event;
        while(window.pollEvent(event))
        {
            if(event.type == sf::Event::Closed)
                window.close();
            if(event.type == sf::Event::KeyPressed)
            {
                if(event.key.control && event.key.code == sf::Keyboard::V)
                    text.setString(sf::Clipboard::getString());
            }
        }

        if(sf::Keyboard::isKeyPressed(sf::Keyboard::Num1))
            sf::Clipboard::setString("This is the first test!");
        if(sf::Keyboard::isKeyPressed(sf::Keyboard::Num2))
            sf::Clipboard::setString(L"Hello world, Καλημέρα κόσμε");

        window.clear();
        window.draw(text);
        window.display();
    }

    return 0;
}

```

mantognini commented on src/SFML/Window/Unix/ClipboardImpl.cpp:
```diff
@@ -0,0 +1,261 @@
+namespace
+{
+////////////////////////////////////////////////////////////
+	void initClipboard();
```
Note: we don't indent on the first namespace.

ricanteja replied:
Noted. Thanks

mantognini commented:
In #715 an example was added as well. Maybe we keep it/do the same? Opinions on that?

eXpl0it3r commented:
Sounds like a good idea.

ricanteja commented:
Thank you! Will do before the end of tomorrow! I'm sorry guys I've been so inactive but my schedule has been crazy lately.

ricanteja commented:
Implemented the requested changes. Also I forgot to note this but in the example I added, I removed any utf8 characters which were not supported by the sansation.ttf font. Didn't want to add another font that might support Greek etc, for consistency sake.

LaurentGomila commented on examples/clipboard/Clipboard.cpp:
```diff
@@ -0,0 +1,79 @@
+
```
Do we really need an example for such a simple thing? Examples are rather demonstrating the use of one or more whole modules, or complex classes.

MarioLiebisch replied:
Yeah, I agree. Better expand the example in `Clipboard.hpp` a bit.

TankOs replied:
I wouldn't mind having examples that even cover small things, but merging it into some bigger example would be nice, yeah.

LaurentGomila commented on examples/clipboard/Clipboard.cpp:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    // Setup strings for later use
+    sf::String str_a = "Hello World!";
```
Variable names such as `str_a` do not follow our naming conventions (should be camelCase).

LaurentGomila commented on include/SFML/Window/Clipboard.hpp:
```diff
@@ -43,27 +43,27 @@ class SFML_WINDOW_API Clipboard
+    ///
+    /// This function returns the content of the clipboard
+    /// as a string.
+    ///
```
The documentation should be more detailed, for example what happens if what's inside the clipboard is not text data (ie filename(s), image data, etc.)? Maybe we should handle that?

MarioLiebisch replied:
At least on Windows there's some automatic conversion going on (and programs can intercept pasting to provide fitting data as well), but I'd at least add copying/pasting images, since it's one core part of SFML. Only problem i see is this adding a dependency back to sfml-graphics? Or should it just return the raw pixels in a vector?

ricanteja replied:
I haven't added support for other forms of data because I haven't thought of a proper API to implement it and also ensure it behaves the same on all of SFML's supported platforms. I don't think images can be copied to the clipboard on Android; maybe on newer versions? I don't know, my phone is old (4.4.2).

LaurentGomila commented on include/SFML/Window/Clipboard.hpp:
```diff
@@ -43,27 +43,27 @@ class SFML_WINDOW_API Clipboard
+    /// This function sets the content of the clipboard to a
+    /// string
+    ///
+    /// \return Current position of the mouse
```
Bad copy and paste :)

LaurentGomila commented on src/SFML/Window/Win32/ClipboardImpl.hpp:
```diff
@@ -44,27 +44,27 @@ class ClipboardImpl
+    /// This function sets the content of the clipboard to a
+    /// string
+    ///
+    /// \return Current position of the mouse
```
Bad copy and paste :)

ricanteja commented:
Ready for another review.

mantognini commented:
Let's see if it builds. :-)

SFMLBuildMaster: Build this please.

I've removed my review as it doesn't apply anymore. Thanks for the changes! I'll try to put together something for the macOS impl soonish. In the meantime, this branch needs to be rebased onto master, and ideally the commits should be squased together into a minimal number of clean commits.

ricanteja commented:
Missing the Android and OSX implementations. Everything else builds.

mantognini commented on include/SFML/Window/Clipboard.hpp:
```diff
@@ -0,0 +1,108 @@
+    /// This function sets the content of the clipboard as a
+    /// string.
+    ///
+    /// \param sf::String containing the data to be sent 
```
While working on the OS X impl, I noticed this warning:

```
In file included from SFML/src/SFML/Window/Clipboard.cpp:28:
SFML/include/SFML/Window/Clipboard.hpp:63:16: warning: parameter 'sf::String' not found in the function declaration [-Wdocumentation]
    /// \param sf::String containing the data to be sent 
               ^~~~~~~~~~
SFML/include/SFML/Window/Clipboard.hpp:63:16: note: did you mean 'text'?
    /// \param sf::String containing the data to be sent 
               ^~~~~~~~~~
               text
1 warning generated.
```

Could you add `text` (the param name) after `\param`? Thanks.

mantognini commented on src/SFML/Window/Clipboard.cpp:
```diff
@@ -0,0 +1,48 @@
+////////////////////////////////////////////////////////////
+#include <SFML/Window/Clipboard.hpp>
+#include <SFML/Window/ClipboardImpl.hpp>
+#include <SFML/System/String.hpp>
```
This include is technically not required but doesn't hurt I'd say.

mantognini commented on src/SFML/Window/Unix/ClipboardImpl.hpp:
```diff
@@ -0,0 +1,78 @@
+////////////////////////////////////////////////////////////
+// Headers
+////////////////////////////////////////////////////////////
+#include <SFML/Window/Clipboard.hpp>
```
This include should be removed.

mantognini commented on src/SFML/Window/Win32/ClipboardImpl.hpp:
```diff
@@ -0,0 +1,77 @@
+////////////////////////////////////////////////////////////
+// Headers
+////////////////////////////////////////////////////////////
+#include <SFML/Window/Clipboard.hpp>
```
This one should also be removed.

mantognini commented on src/SFML/Window/Win32/ClipboardImpl.hpp:
```diff
@@ -0,0 +1,77 @@
+    /// This function sets the content of the clipboard as a
+    /// string.
+    ///
+    /// \param sf::String object containing the data to be sent 
```
This should be updated as well.

mantognini commented on src/SFML/Window/Unix/ClipboardImpl.hpp:
```diff
@@ -0,0 +1,78 @@
+    /// This function sets the content of the clipboard as a
+    /// string.
+    ///
+    /// \param sf::String object containing the data to be sent 
```
This should be updated as well.

ricanteja commented:
@mantognini I added the changes to the doc. If I knew how to 'squash' the commits I would but I got no idea. I tried but couldn't get it to work. Sorry about all the spam in the git history! I haven't been using git for a long time you can tell :)

mantognini commented:
It's okay for the squash -- I'll do it myself since I have to add the macOS implementation anyway. Unless someone raises some concerns about the X11 / Win32 impls, say before Wednesday, I'll take care of creating a new PR with your work.

mantognini commented:
SFMLBuildMaster: Build this please.

mantognini commented:
Superseded by #1221.
